java/ql/src/semmle/code/java/dataflow/internal/ContainerFlow.qll

```diff
@@ -88,6 +91,290 @@ class ContainerType extends RefType {
+        "java.util;Collections;false;unmodifiableMap;(Map);;MapKey of Argument[0];MapKey of ReturnValue;value",
+        "java.util;Collections;false;unmodifiableMap;(Map);;MapValue of Argument[0];MapValue of ReturnValue;value",
+        "java.util;Collections;false;unmodifiableList;(List);;Element of Argument[0];Element of ReturnValue;value",
+        "java.util;Collections;false;unmodifiableNavigableSet;(NavigableSet);;Element of Argument[0];Element of ReturnValue;value",
+        "java.util;Collections;false;unmodifiableSortedSet;(SortedSet);;Element of Argument[0];Element of ReturnValue;value",
+        "java.util;Collections;false;unmodifiableSet;(Set);;Element of Argument[0];Element of ReturnValue;value",
+        "java.util;Collections;false;unmodifiableCollection;(Collection);;Element of Argument[0];Element of ReturnValue;value",
+        "java.util;Collections;false;max;(Collection,Comparator);;Element of Argument[0];ReturnValue;value",
+        "java.util;Collections;false;max;(Collection);;Element of Argument[0];ReturnValue;value",
+        "java.util;Collections;false;min;(Collection,Comparator);;Element of Argument[0];ReturnValue;value",
+        "java.util;Collections;false;min;(Collection);;Element of Argument[0];ReturnValue;value",
+        "java.util;Arrays;false;fill;(Object[],int,int,Object);;Argument[3];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(Object[],Object);;Argument[1];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(float[],int,int,float);;Argument[3];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(float[],float);;Argument[1];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(double[],int,int,double);;Argument[3];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(double[],double);;Argument[1];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(boolean[],int,int,boolean);;Argument[3];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(boolean[],boolean);;Argument[1];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(byte[],int,int,byte);;Argument[3];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(byte[],byte);;Argument[1];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(char[],int,int,char);;Argument[3];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(char[],char);;Argument[1];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(short[],int,int,short);;Argument[3];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(short[],short);;Argument[1];ArrayElement of Argument[0];value",
+        "java.util;Arrays;false;fill;(int[],int,int,int);;Argument[3];ArrayElement of Argument[0];value",
```

Comment:
I note lots of primitive propagation steps that are ruled out elsewhere?

Reply:
Where are they ruled out?

Reply:
I thought these would be ruled out by the numeric-type exclusions discussed in the other thread, but perhaps not?

Reply:
No. The bulk of this work is about adding value-flow through arrays and collections by modelling the steps as proper read- and store steps (aka as field flow). This means that the PR removes those steps from the default set of taint steps, and instead adds them as general read- and store steps that use the synthetic field `ArrayElement`. The array read step with the type exclusions in `TaintTrackingUtil.qll` is a taint step and not a read step.

Reply:
👍